presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,560 @@
+                    source,
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(unnestNode.getReplicateVariables()).addAll(expressions.values()).build(),
+                    unnestNode.getUnnestVariables(),
+                    unnestNode.getOrdinalityVariable());
+            return Result.ofPlanNode(newUnnest);
+        }
+    }
+
+    private static PlanNode createProjectBelow(PlanNode planNode, Assignments dereferences, PlanNodeIdAllocator idAllocator)
+    {
+        if (dereferences.isEmpty()) {
+            return planNode;
+        }
+        return new ProjectNode(idAllocator.getNextId(), planNode, Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(planNode.getOutputVariables())).putAll(dereferences).build());
+    }
+
+    private static class DereferenceReplacer
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<DereferenceExpression, VariableReferenceExpression> expressions;
+
+        DereferenceReplacer(Map<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            this.expressions = requireNonNull(expressions, "expressions is null");
+        }
+
+        @Override
+        public Expression rewriteDereferenceExpression(DereferenceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (expressions.containsKey(node)) {
+                return new SymbolReference(expressions.get(node).getName());
+            }
+            return treeRewriter.defaultRewrite(node, context);
+        }
+    }
+
+    private static List<DereferenceExpression> extractDereferenceExpressions(Expression expression)
+    {
+        ImmutableList.Builder<DereferenceExpression> builder = ImmutableList.builder();
+        new DefaultExpressionTraversalVisitor<Void, ImmutableList.Builder<DereferenceExpression>>()
+        {
+            @Override
+            protected Void visitDereferenceExpression(DereferenceExpression node, ImmutableList.Builder<DereferenceExpression> context)
+            {
+                context.add(node);
+                return null;
+            }
+        }.process(expression, builder);
+        return builder.build();
+    }
+
+    private static Map<DereferenceExpression, VariableReferenceExpression> getDereferenceSymbolMap(Collection<RowExpression> expressions, Context context, Metadata metadata, SqlParser sqlParser)
+    {
+        Set<DereferenceExpression> dereferences = expressions.stream()
+                .filter(OriginalExpressionUtils::isExpression)
+                .map(OriginalExpressionUtils::castToExpression)
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .filter(PushDownDereferences::validPushDown)
+                .collect(toImmutableSet());
+
+        // TODO DereferenceExpression Base will cause unnecessary rewritten of same expression.
```

Comment:
I'm not sure I understand this TODO and the logic in the following if statement. Looks like [msg.foo, msg.foo.bar] can is not being processed at all. Is this intentional? Shouldn't this case result in pushing down msg.foo?